### DISCUSSION TITLE
Reality Mesh is located under Bing Map

### DISCUSSION THREAD

**Discussion by namsancow** [iTwin/itwinjs-core#330] [Q&A] [answered]
I have tested to attach the Reality Mesh with Bing Map in imodeljs. 
Although I changed the geographic system, the Reality Mesh is always located under surface of Bing Map.
And the distance between the Reality Mesh and the surface of Bing Map is dozen of kilometers.

**Comment by RBBentley** [comment]
Can you provide an example -- Do you have terrain enabled?

**Reply by namsancow** [reply]
> Can you provide an example -- Do you have terrain enabled?

Many thanks to your kind response.

As I mentioned before, the Reality Mesh is not overlapped exactly on the surface of Bing Map.

I applied GCS as EPSG3857 for Bing Map and Google Map while making the Reality Mesh,
however, the Reality Mesh is always positioned over dozen of kilometers from the Bing Map surface(GCS_EPSG3857_RealityMesh_over_BingMap_01.png). 

Although the z-axis position of the Reality Mesh is not correct, the x-axis, y-axis positions is correct(GCS_EPSG3857_RealityMesh_over_BingMap_02.png).

You can check this issue from the attached zip file.

[GCS_EPSG3857.zip](https://github.com/imodeljs/imodeljs/files/5656986/GCS_EPSG3857.zip)

![GCS_EPSG3857_RealityMesh_Capture](https://user-images.githubusercontent.com/22971147/101440925-3e99b500-395b-11eb-921f-7a4262cc9a16.png)

![GCS_EPSG3857_RealityMesh_over_BingMap_01](https://user-images.githubusercontent.com/22971147/101440992-638e2800-395b-11eb-87e8-b9d91e590f29.png)

![GCS_EPSG3857_RealityMesh_over_BingMap_02](https://user-images.githubusercontent.com/22971147/101441828-5114ee00-395d-11eb-9ca5-6b19afc54a39.png)

**Reply by AlainRobertAtBentley** [reply]
@namsamcow Thank you for the dataset. We have started studying it to locate the problem and will get back to you as soon as possible.

**Reply by RBBentley** [reply]
Thanks Alain --  I believe iModel.js is displaying this reality model in the correct location.  It is just mislocated (I verified by looking at it in CesiuM) .   - As it came from scalable mesh, perhaps you can help get it located in correct location.

**Reply by AlainRobertAtBentley** [reply]
@namsancow The model appears to have a flaw in the transformation that converts from its native coordinate system(EPSG:3857) to ECEF. This transformation is used by imodeljs for the model positionning.
On the other hand the model appears at the proper position in MicroStation when opened in a DGN model using EPSG:3857 (The transform is then not necessary). Elevation is then between something like 19 to 25 meters. The header in the dataset indicates the data has been generated using Bentley Context Capture version 10.16.0.75. Can you confirm it is the case?
Since imodelJS displays the model correctly according to the content can we take this out of the imodelJS github? Could you file a Service Request concerning Context Capture. You can mention my name (Alain Robert) in the Service Request as well as the present web page. This request will be forwarded to Context Capture support personnel who will contact you about the details of the model creation.
In the meantime I have contacted the concerned developpers and we will continue to investigate how this dataset was produced and how it can be corrected.
Regards

**Comment by namsancow** [comment]
Dear RBBentley and AlainRobertAtBentley

In the meantime, I've been studying the issues that the Reality Mesh on the BingMap looks strange.

First of all, if the height of the geographic coordinate system of the Reality Mesh is set to -2m, and if you look as the "Birds Eye" view, the Reality Mesh is overlapped to the BingMap and looks normal. If you set the height to 0m, it looks a bit floating, but the Reality Mesh still looks somewhat close to the BingMap. For more information, see the attached the files of BirdseyeV2WithLabels.jpg and BirdseyeV2WithLabels.mp4.

However, if the geographic coordinate system height of the real model is set to -2m, and the "Aerial With Label" view shows that the Reality Mesh does not look normal on the BingMap and appears in the wrong place. Setting the height to 0 m results in the same abnormality. In the custom viewer made with imodeljs source, there also appears to be a Reality Mesh 20km above, and in the imodel "Desktop-Starter", it appears and disappears buried in an BingMap. For more information, you can check the attached the files of AerialWithLabel.jpg and AerialWithLabel.mp4.

I would appreciate it if you let me know if you have a seperate way to solve the abnormality.
![BirdseyeV2WithLabels](https://user-images.githubusercontent.com/22971147/102749286-196b6480-43a7-11eb-92b1-31eb17093b66.jpg)

https://user-images.githubusercontent.com/22971147/102749293-1e301880-43a7-11eb-995b-339212d031ab.mp4

![AerialWithLabel](https://user-images.githubusercontent.com/22971147/102749308-27b98080-43a7-11eb-9ff9-42fc275fbbf3.jpg)

https://user-images.githubusercontent.com/22971147/102749309-2a1bda80-43a7-11eb-9643-623ce9fd65a6.mp4

**Comment by RBBentley** [comment, marked as answer]
I'd have to look at your model myself - Are you viewing with terrain enabled? -- If so (and your reality model is located accurately) then I would expect the model to be located correctly.   If you are viewing without terrain then the default is to not depth buffer the map such that the BIM geometry should always be on top.  I am currently working on clipping the terrain to geometry model boundaries so that the two will not ever interfere which may be helpful.   Can you provide a testcase?  - Thanks.

**Reply by namsancow** [reply]
> I'd have to look at your model myself - Are you viewing with terrain enabled? -- If so (and your reality model is located accurately) then I would expect the model to be located correctly. If you are viewing without terrain then the default is to not depth buffer the map such that the BIM geometry should always be on top. I am currently working on clipping the terrain to geometry model boundaries so that the two will not ever interfere which may be helpful. Can you provide a testcase? - Thanks.

Dear RBBentley
Many thanks to your kind answer.
As your recommendation, I can overlap the my Reality Mesh on BingMap normally!!!
Problems such as the Reality Mesh was floating on the BingMap, located under the BingMap, or visible or invisible have all been resolved.
Imodeljs is the best!!!
